File: app/src/main/java/com/aayush/filmipedia/view/fragment/SearchFragment.java
package com.aayush.filmipedia.view.fragment;

import android.app.SearchManager;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.aayush.filmipedia.FilmipediaApplication;
import com.aayush.filmipedia.R;
import com.aayush.filmipedia.model.Movie;
import com.aayush.filmipedia.model.PersonResult;
import com.aayush.filmipedia.util.adapter.MovieResultsAdapter;
import com.aayush.filmipedia.util.adapter.PersonAdapter;
import com.aayush.filmipedia.viewmodel.MovieResultViewModel;
import com.aayush.filmipedia.viewmodel.PersonViewModel;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatRadioButton;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class SearchFragment extends Fragment {
    @BindView(R.id.btn_movie)     AppCompatRadioButton movieButton;
    @BindView(R.id.btn_person)    AppCompatRadioButton personButton;
    @BindView(R.id.search_view)   SearchView searchView;
    @BindView(R.id.movie_results) RecyclerView recyclerView;

    private MovieResultViewModel movieResultViewModel;
    private MovieResultsAdapter movieResultsAdapter;
    private List<Movie> movieList;
    private PersonViewModel personViewModel;
    private PersonAdapter personAdapter;
    private List<PersonResult> personResultList;

    private Unbinder unbinder;

    public SearchFragment() {}

    public static SearchFragment newInstance() {
        return new SearchFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_search, container, false);
        unbinder = ButterKnife.bind(this, view);

        initViews();

        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    private void initViews() {
        movieButton.setOnClickListener(v -> {
            personButton.setChecked(!movieButton.isChecked());
            movieResultViewModel = new MovieResultViewModel(FilmipediaApplication.create(getContext()),
                    "");
            movieList = new ArrayList<>();
            movieResultsAdapter = new MovieResultsAdapter(getContext(), movieList);
            movieResultViewModel.getMovieLiveData().observe(SearchFragment.this,
                    pagedList -> movieResultsAdapter.submitList(pagedList));
            movieResultViewModel.getNetworkState().observe(SearchFragment.this,
                    networkState -> movieResultsAdapter.setNetworkState(networkState));
            recyclerView.setAdapter(movieResultsAdapter);
        });
        personButton.setOnClickListener(v -> {
            movieButton.setChecked(!personButton.isChecked());
            personViewModel = new PersonViewModel(FilmipediaApplication.create(getContext()),
                    "");
            personResultList = new ArrayList<>();
            personAdapter = new PersonAdapter(getContext(), personResultList);
            personViewModel.getPersonLiveData().observe(SearchFragment.this,
                    pagedList -> personAdapter.submitList(pagedList));
            personViewModel.getNetworkState().observe(SearchFragment.this,
                    networkState -> personAdapter.setNetworkState(networkState));
            recyclerView.setAdapter(personAdapter);
        });

        movieResultViewModel = new MovieResultViewModel(FilmipediaApplication.create(getContext()),
                "");
        movieList = new ArrayList<>();
        movieResultsAdapter = new MovieResultsAdapter(getContext(), movieList);
        personViewModel = new PersonViewModel(FilmipediaApplication.create(getContext()),
                "");
        personResultList = new ArrayList<>();
        personAdapter = new PersonAdapter(getContext(), personResultList);

        initSearch();

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        if (movieButton.isChecked()) {
            movieResultViewModel.getMovieLiveData().observe(SearchFragment.this,
                    pagedList -> movieResultsAdapter.submitList(pagedList));
            movieResultViewModel.getNetworkState().observe(SearchFragment.this,
                    networkState -> movieResultsAdapter.setNetworkState(networkState));
            recyclerView.setAdapter(movieResultsAdapter);
        }
        else {
            personViewModel.getPersonLiveData().observe(SearchFragment.this,
                    pagedList -> personAdapter.submitList(pagedList));
            personViewModel.getNetworkState().observe(SearchFragment.this,
                    networkState -> personAdapter.setNetworkState(networkState));
            recyclerView.setAdapter(personAdapter);
        }
    }

    private void initSearch() {
        SearchManager searchManager = (SearchManager) Objects.requireNonNull(getActivity())
                .getSystemService(Context.SEARCH_SERVICE);
        searchView.setSearchableInfo(searchManager
                .getSearchableInfo(getActivity()
                        .getComponentName()));
        searchView.setIconifiedByDefault(false);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                searchView.clearFocus();
                performSearch(query);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                if (newText.length() > 3) {
                    performSearch(newText);
                }
                return false;
            }
        });
    }

    private void performSearch(String query) {
        if (movieButton.isChecked()) {
            if (movieResultsAdapter != null) {
                movieResultsAdapter.clear();
            }
            else {
                movieList = new ArrayList<>();
                movieResultsAdapter = new MovieResultsAdapter(getContext(), movieList);
            }
            movieResultViewModel = new MovieResultViewModel(FilmipediaApplication.create(getContext()),
                    query);
            movieResultViewModel.getMovieLiveData().observe(this,
                    movies -> movieResultsAdapter.submitList(movies));
            movieResultViewModel.getNetworkState().observe(this,
                    networkState -> movieResultsAdapter.setNetworkState(networkState));
        }
        else {
            if (personAdapter != null) {
                personAdapter.clear();
            }
            else {
                personResultList = new ArrayList<>();
                personAdapter = new PersonAdapter(getContext(), personResultList);
            }
            personViewModel = new PersonViewModel(FilmipediaApplication.create(getContext()),
                    query);
            personViewModel.getPersonLiveData().observe(this,
                    personResults -> personAdapter.submitList(personResults));
            personViewModel.getNetworkState().observe(this,
                    networkState -> personAdapter.setNetworkState(networkState));
        }
    }
}

File: app/src/main/java/com/aayush/filmipedia/view/fragment/SearchFragment.kt
package com.aayush.filmipedia.view.fragment

import android.app.SearchManager
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.aayush.filmipedia.FilmipediaApplication
import com.aayush.filmipedia.R
import com.aayush.filmipedia.model.Movie
import com.aayush.filmipedia.model.PersonResult
import com.aayush.filmipedia.util.adapter.MovieResultsAdapter
import com.aayush.filmipedia.util.adapter.PersonAdapter
import com.aayush.filmipedia.viewmodel.MovieResultViewModel
import com.aayush.filmipedia.viewmodel.PersonViewModel
import kotlinx.android.synthetic.main.fragment_search.*

class SearchFragment: Fragment() {
    private lateinit var movieResultViewModel: MovieResultViewModel
    private lateinit var movieResultsAdapter: MovieResultsAdapter
    private lateinit var movieList: MutableList<Movie>
    private lateinit var personViewModel: PersonViewModel
    private lateinit var personAdapter: PersonAdapter
    private lateinit var personResultList: MutableList<PersonResult>

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_search, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initViews()
    }

    private fun initViews() {
        btn_movie.setOnClickListener {
            btn_person.isChecked = !btn_movie.isChecked
            movieList = arrayListOf()
            movieResultViewModel = MovieResultViewModel(
                FilmipediaApplication.getApplication(context!!),
                ""
            ).apply {
                this.movieLiveData.observe(this@SearchFragment, Observer(movieResultsAdapter::submitList))
                this.networkState.observe(this@SearchFragment, Observer(movieResultsAdapter::setNetworkState))
            }
            movieResultsAdapter = MovieResultsAdapter(context!!, movieList).apply { movie_results.adapter = this }
        }
        btn_person.setOnClickListener {
            btn_movie.isChecked = !btn_person.isChecked
            personViewModel = PersonViewModel(FilmipediaApplication.getApplication(context!!), "")
            personResultList = arrayListOf()
            personAdapter = PersonAdapter(context!!, personResultList)
            personViewModel.apply {
                personLiveData.observe(this@SearchFragment, Observer(personAdapter::submitList))
                networkState.observe(this@SearchFragment, Observer(personAdapter::setNetworkState))
            }
            movie_results.adapter = personAdapter
        }

        movieList = arrayListOf()
        movieResultViewModel = MovieResultViewModel(FilmipediaApplication.getApplication(context!!), "")
        movieResultsAdapter = MovieResultsAdapter(context!!, movieList)
        personViewModel = PersonViewModel(FilmipediaApplication.getApplication(context!!), "")
        personResultList = arrayListOf()
        personAdapter = PersonAdapter(context!!, personResultList)

        initSearch()

        movie_results.layoutManager = LinearLayoutManager(context!!)

        if (btn_movie.isChecked) {
            movieResultViewModel.apply {
                movieLiveData.observe(this@SearchFragment, Observer(movieResultsAdapter::submitList))
                networkState.observe(this@SearchFragment, Observer(movieResultsAdapter::setNetworkState))
            }
            movie_results.adapter = movieResultsAdapter
        } else {
            personViewModel.apply {
                personLiveData.observe(this@SearchFragment, Observer(personAdapter::submitList))
                networkState.observe(this@SearchFragment, Observer(personAdapter::setNetworkState))
            }
            movie_results.adapter = personAdapter
        }
    }

    private fun initSearch() {
        val searchManager = activity?.getSystemService(Context.SEARCH_SERVICE) as SearchManager
        search_view.apply {
            setSearchableInfo(searchManager.getSearchableInfo(activity!!.componentName))
            setIconifiedByDefault(false)
            setOnQueryTextListener(object: SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    this@apply.clearFocus()
                    performSearch(query)
                    return true
                }

                override fun onQueryTextChange(newText: String): Boolean {
                    if (newText.length > 3) {
                        performSearch(newText)
                    }
                    return false
                }
            })
        }
    }

    private fun performSearch(query: String) {
        if (btn_movie.isChecked) {
            movieResultsAdapter.clear()
            movieResultViewModel = MovieResultViewModel(FilmipediaApplication.getApplication(context!!), query).apply { 
                this.movieLiveData.observe(this@SearchFragment, Observer(movieResultsAdapter::submitList))
                this.networkState.observe(this@SearchFragment, Observer(movieResultsAdapter::setNetworkState))
            }
        } else {
            personAdapter.clear()
            personViewModel = PersonViewModel(FilmipediaApplication.getApplication(context!!), query).apply {
                this.personLiveData.observe(this@SearchFragment, Observer(personAdapter::submitList))
                this.networkState.observe(this@SearchFragment, Observer(personAdapter::setNetworkState))   
            }
        }
    }

    companion object {
        fun newInstance() = SearchFragment()
    }
}
